src/init/starship.zsh

```diff
@@ -3,93 +3,90 @@
+autoload is-at-least
+zmodload zsh/parameter # Needed to access jobstates variable for STARSHIP_JOBS_COUNT
 
-# Defines a function `__starship_get_time` that sets the time since epoch in millis in STARSHIP_CAPTURED_TIME.
-if [[ $ZSH_VERSION == ([1-4]*) ]]; then
-    # ZSH <= 5; Does not have a built-in variable so we will rely on Starship's inbuilt time function.
-    __starship_get_time() {
-        STARSHIP_CAPTURED_TIME=$(::STARSHIP:: time)
-    }
-else
+if is-at-least 4.3.13; then
     zmodload zsh/datetime
     zmodload zsh/mathfunc
-    __starship_get_time() {
-        (( STARSHIP_CAPTURED_TIME = int(rint(EPOCHREALTIME * 1000)) ))
+    __starship_time() {
+        echo $(( int(rint(EPOCHREALTIME * 1000)) ))
+    }
+else
+    __starship_time() {
+        ::STARSHIP:: time
     }
 fi
 
 # Will be run before every prompt draw
-starship_precmd() {
+__starship_precmd() {
     # Save the status, because commands in this pipeline will change $?
     STARSHIP_CMD_STATUS=$?
 
     # Compute cmd_duration, if we have a time to consume, otherwise clear the
     # previous duration
-    if (( ${+STARSHIP_START_TIME} )); then
-        __starship_get_time && (( STARSHIP_DURATION = STARSHIP_CAPTURED_TIME - STARSHIP_START_TIME ))
-        unset STARSHIP_START_TIME
+    if (( ${+STARSHIP_CMD_START_TIME} )); then
+        STARSHIP_CMD_DURATION=$(( $(__starship_time) - STARSHIP_CMD_START_TIME ))
+        unset STARSHIP_CMD_START_TIME
     else
-        unset STARSHIP_DURATION
+        unset STARSHIP_CMD_DURATION
     fi
 
     # Use length of jobstates array as number of jobs. Expansion fails inside
     # quotes so we set it here and then use the value later on.
     STARSHIP_JOBS_COUNT=${#jobstates}
 }
-starship_preexec() {
-    __starship_get_time && STARSHIP_START_TIME=$STARSHIP_CAPTURED_TIME
+__starship_preexec() {
+    STARSHIP_CMD_START_TIME="$(__starship_time)"
 }
 
 # If precmd/preexec arrays are not already set, set them. If we don't do this,
-# the code to detect whether starship_precmd is already in precmd_functions will
-# fail because the array doesn't exist (and same for starship_preexec)
+# the code to detect whether __starship_precmd is already in precmd_functions will
+# fail because the array doesn't exist (and same for __starship_preexec)
 (( ! ${+precmd_functions} )) && precmd_functions=()
 (( ! ${+preexec_functions} )) && preexec_functions=()
 
 # If starship precmd/preexec functions are already hooked, don't double-hook them
 # to avoid unnecessary performance degradation in nested shells
-if [[ -z ${precmd_functions[(re)starship_precmd]} ]]; then
-    precmd_functions+=(starship_precmd)
-fi
-if [[ -z ${preexec_function[(re)starship_preexec]} ]]; then
-    preexec_functions+=(starship_preexec)
-fi
+(( ! ${precmd_functions[(Ie)__starship_precmd]} )) && precmd_functions+=(__starship_precmd)
+(( ! ${preexec_functions[(Ie)__starship_preexec]} )) && preexec_functions+=(__starship_preexec)
```

Comment:
Maybe it can use [`add-zsh-hook` module](https://zsh.sourceforge.io/Doc/Release/User-Contributions.html#Manipulating-Hook-Functions) of zsh, this method is also recommended by [zsh official theme document](https://zsh.sourceforge.io/Doc/Release/User-Contributions.html#Writing-Themes). 
Using this method is good for implementation of https://github.com/starship/starship/issues/1994 as well.